Generate AST when importing a cpp file

bricknerb opened the pull request:
Ignore the AST and support a single Cpp import, for now.
Report cpp compilation errors and warnings.
Part of #4666

danakj commented on toolchain/check/check_unit.cpp:
```diff
@@ -29,9 +35,11 @@ static auto GetImportedIRCount(UnitAndImports* unit_and_imports) -> int {
                      llvm::raw_ostream* vlog_stream)
     : unit_and_imports_(unit_and_imports),
       total_ir_count_(total_ir_count),
+      fs_(fs),
```
```suggestion
      fs_(std::move(fs)),
```
(avoid refcount churn)

bricknerb replied:
Done.

danakj commented on toolchain/check/check_unit.cpp:
```diff
@@ -325,6 +334,37 @@ auto CheckUnit::ImportOtherPackages(SemIR::TypeId namespace_type_id) -> void {
+
+    auto file = fs_->openFileForRead(cpp_file_path);
+    if (!file) {
+      CARBON_DIAGNOSTIC(CppInteropFileNotFound, Error,
```
```suggestion
      CARBON_DIAGNOSTIC(CppInteropCantReadFile, Error,
```
The text and the name say different things, they could be more similar?

bricknerb replied:
Done.

danakj commented on toolchain/check/import_cpp.cpp:
```diff
@@ -0,0 +1,57 @@
+  if (num_errors > 0) {
+    // TODO: Remove the warnings part when there are no warnings.
+    CARBON_DIAGNOSTIC(
+        CppInteropCodeErrors, Error,
```
```suggestion
        CppInteropParseError, Error,
```

bricknerb replied:
Done.

danakj commented on toolchain/check/import_cpp.cpp:
```diff
@@ -0,0 +1,57 @@
+    context.emitter().Emit(loc, CppInteropCodeErrors, num_errors, num_warnings,
+                           file_path.str(), diagnostics_str);
+  } else if (num_warnings > 0) {
+    CARBON_DIAGNOSTIC(CppInteropCodeWarnings, Warning,
```
```suggestion
    CARBON_DIAGNOSTIC(CppInteropParseWarning, Warning,
```

bricknerb replied:
Done.

bricknerb committed:
Generate AST when importing cpp files
Ignore the AST, for now.
Report cpp compilation errors and warnings.
Part of #4666

jonmeow commented on toolchain/check/check.cpp:
```diff
@@ -66,11 +66,20 @@ static auto TrackImport(Map<ImportKey, UnitAndImports*>& api_map,
-  bool is_explicit_main = import_key.first == MainPackageName;
+  bool is_explicit_main = import_package_name == MainPackageName;
+
+  bool is_cpp =
+      import_package_name == CppPackageName && !import_library_name.empty();
+
+  if (is_cpp) {
+    unit_info.cpp_imports.push_back(import);
+    return;
+  }
```
Would it make sense to move this above `is_explicit_main` since `is_explicit_main` isn't used in this path? Also, does `is_cpp` need the local or is it only used here? (note for contrast, `is_explicit_main` is referenced in a few spots)

bricknerb replied:
Done and removed `is_cpp`.

jonmeow commented on toolchain/check/check.cpp:
```diff
@@ -66,11 +66,20 @@ static auto TrackImport(Map<ImportKey, UnitAndImports*>& api_map,
-  bool is_explicit_main = import_key.first == MainPackageName;
+  bool is_explicit_main = import_package_name == MainPackageName;
+
+  bool is_cpp =
+      import_package_name == CppPackageName && !import_library_name.empty();
```
Can you add some tests for things like:
- `import package Cpp;` -- should probably be complaining that `Cpp` needs a library, might complain that there's no `api` file right now.
- `import package Cpp library "";` - "file not found" seems fine
- `import package Cpp library "\"foo.h\"";` - ???

bricknerb replied:
Added, but without the `package`, since `import` doesn't seem to support `import package`.

danakj commented on toolchain/check/testdata/interop/cpp/no_prelude/file_not_found.carbon:
```diff
@@ -0,0 +1,25 @@
+
+library "[[@TEST_NAME]]";
+
+// CHECK:STDERR: fail_cpp_file_not_found.carbon:[[@LINE+3]]:1: error: file 'not_found.h' couldn't be opened for reading: No such file or directory [CppInteropFileNotFound]
```
I think the "No such file or directory" reason could move to a note here

bricknerb replied:
Done.

danakj commented on toolchain/diagnostics/coverage_test.cpp:
```diff
@@ -41,10 +42,9 @@ constexpr DiagnosticKind UntestedDiagnosticKinds[] = {
-                            llvm::ArrayRef(DiagnosticKinds),
-                            llvm::ArrayRef(UntestedDiagnosticKinds));
+  Testing::TestKindCoverage(
+      absl::GetFlag(FLAGS_testdata_manifest), R"( \[([A-Z]\w*)\]$)",
```
This regex will match most anything now, is it still effectively testing something?

bricknerb replied:
I agree it's quite wider, but seems to work fine for now.
I can think of alternatives but we might want to adjust it when we have a new use case that doesn't fit.
Let me know what you think.

jonmeow replied:
I'm similarly concerned that this is more broad than it needs to be. Does this work if you change it back to only looking in diagnostic output, e.g.:

```suggestion
      absl::GetFlag(FLAGS_testdata_manifest), R"(^ *// CHECK:STDERR: .* \[(\w+)\]$)",
```

Note, in general I think it's good to figure out ways to make regexes restrictive. The check as it stands would match something like an array in source code, so a test like:

```
var x: [N]
```

bricknerb replied:
Makes sense, thanks, done.

danakj reviewed (commented):
a couple suggestions and questions

jonmeow commented on toolchain/check/check_unit.cpp:
```diff
@@ -325,6 +334,37 @@ auto CheckUnit::ImportOtherPackages(SemIR::TypeId namespace_type_id) -> void {
+                    cpp_file_path.str(), buffer.getError().message());
+      continue;
+    }
+    ImportCppFile(context_, import.node_id, cpp_file_path,
+                  buffer.get()->getBuffer());
```
Versus loading each C++ file individually, had you considered generating a file like:

```
  std::string content;
  for (const auto& import : unit_and_imports_->cpp_imports) {
    llvm::StringRef cpp_file_path =
        unit_and_imports_->unit->value_stores->string_literal_values().Get(
            import.library_id);
    content += llvm::formatv("#include \"{0}\"\n", cpp_file_path);
  }
  ImportCppFile(context_, import.node_id, carbon_file_name + '.interop.cpp", content);
```

One particular consequence of that kind of approach is that if you have something like C++ headers that interact (e.g. one #define's a value, the next has a #ifdef on it) they'd work just like they would in C++. Or if there are overloaded functions, it lets C++ decide how to handle them. Thoughts?

(note, my example uses a simple `#include`, but down the road we'll want to think about how to work with C++ modules too)

danakj replied:
That's a good idea. It'll also be faster as it'll have to build/walk an AST once instead of N times, and only has to go through transitive includes once.

bricknerb replied:
Thanks, you're absolutely right that multiple Cpp imports should probably behave different.
I'm a bit concerned about how to support diagnostics well when we set have multiple Cpp imports and the impact that one import has on another might be confusing, though could be necessary.
I've decided to delay this by putting multiple Cpp imports out of scope in this PR, and now explicitly generate an error for this use case.

jonmeow replied:
To be sure, I think this will block support of multiple imports, which we will need. Are you sure you'd rather not make it work first, then sort out improving diagnostics later?

bricknerb replied:
What do you mean by "block support"?
We will support this, just not in this PR.

jonmeow reviewed (commented):
This is great!

I'm looking at this, and thought of a lateral issue... We're fuzz testing, and as soon as the fuzzer discovers it can write C++ imports we'll end up fuzz testing Clang, and Clang is not crash-resilient. The driver already has a `DriverEnv::fuzzing` flag, can you have it lead to an error when set and importing from Cpp?

bricknerb committed:
Address review comments:
1. Change diagnostic ids.
2. Split file read error into a note.
3. Added bad import Cpp tests.
4. Remove `is_cpp` and move the condition higher.

bricknerb committed:
Explicitly unsupport multiple Cpp imports

bricknerb committed:
When using Carbon Driver for Fuzzing, diagnose an error when trying to import Cpp files.

bricknerb commented:
> This is great!
> 
> I'm looking at this, and thought of a lateral issue... We're fuzz testing, and as soon as the fuzzer discovers it can write C++ imports we'll end up fuzz testing Clang, and Clang is not crash-resilient. The driver already has a `DriverEnv::fuzzing` flag, can you have it lead to an error when set and importing from Cpp?

I've added an error for that.

danakj commented on toolchain/check/check.cpp:
```diff
@@ -61,26 +61,31 @@ static auto RenderImportKey(ImportKey import_key) -> std::string {
+    if (fuzzing) {
+      // Clang is not crash-resilient.
+      CARBON_DIAGNOSTIC(CppInteropFuzzing, Error,
+                        "Cpp import is not supported in fuzz testing");
```
I think "Use "cannot" if needed, but try to use phrasing that doesn't require it. Avoid "allowed", "legal", "permitted", "valid", and related wording. For example:" applies here, from https://github.com/carbon-language/carbon-lang/blob/230a8ee598414334e224017e4d87c75061496046/toolchain/docs/diagnostics.md#diagnostic-message-style-guide

```suggestion
                        "Cpp import found during fuzzing");
```

bricknerb replied:
Thanks, done.

danakj commented on toolchain/check/check_unit.cpp:
```diff
@@ -335,34 +335,50 @@ auto CheckUnit::ImportOtherPackages(SemIR::TypeId namespace_type_id) -> void {
+  if (imports.size() >= 2) {
+    // TODO: Support multiple Cpp imports.
+    CARBON_DIAGNOSTIC(CppInteropMultipleImports, Error,
+                      "multiple ({0}) Cpp imports are not supported", int);
```
```suggestion
                      "multiple ({0}) Cpp imports are not yet supported", int);
```

jonmeow replied:
Had you considered using `context.TODO` for this, instead of defining a diagnostic?

bricknerb replied:
Added "yet".
I didn't consider `context.TODO`, thanks for pointing to that! Done.

danakj commented on toolchain/check/testdata/interop/cpp/no_prelude/bad_import.carbon:
```diff
@@ -0,0 +1,96 @@
+
+library "[[@TEST_NAME]]";
+
+// CHECK:STDERR: fail_import_cpp_library_empty.carbon:[[@LINE+4]]:1: error: imported API 'Cpp' not found [ImportNotFound]
```
This is a bit weird, the same diagnostic as `import Cpp;`. I was expecting `file '' not found`. Maybe we can improve this in a followup.

bricknerb replied:
We now diagnose with `Cpp` import missing library

danakj commented on toolchain/check/check_unit.cpp:
```diff
@@ -335,34 +335,50 @@ auto CheckUnit::ImportOtherPackages(SemIR::TypeId namespace_type_id) -> void {
+  auto buffer = file_ref.getBuffer(cpp_file_path);
+  if (!buffer) {
+    CARBON_DIAGNOSTIC(CppInteropFailedAccessingFileBuffer, Error,
+                      "failed accessing buffer of file '{0}': {1}", std::string,
```
The CppInteropCantReadFileNote is much nicer, thanks. Should we move {1} to a note here as well?

bricknerb replied:
We're now use `SourceBuffer` for that, so this is obsolete.

jonmeow commented on toolchain/check/check_unit.cpp:
```diff
@@ -325,6 +334,53 @@ auto CheckUnit::ImportOtherPackages(SemIR::TypeId namespace_type_id) -> void {
+    return;
+  }
+
+  const auto& import = imports.front();
+  llvm::StringRef cpp_file_path =
+      unit_and_imports_->unit->value_stores->string_literal_values().Get(
+          import.library_id);
+
+  auto file = fs_->openFileForRead(cpp_file_path);
+  if (!file) {
+    CARBON_DIAGNOSTIC(CppInteropCantReadFile, Error,
+                      "file '{0}' couldn't be opened for reading", std::string);
+    CARBON_DIAGNOSTIC(CppInteropCantReadFileNote, Note, "File read error: {0}",
+                      std::string);
+    emitter_.Build(import.node_id, CppInteropCantReadFile, cpp_file_path.str())
+        .Note(import.node_id, CppInteropCantReadFileNote,
+              file.getError().message())
+        .Emit();
+    return;
+  }
+
+  llvm::vfs::File& file_ref = *file.get();
+  auto buffer = file_ref.getBuffer(cpp_file_path);
+  if (!buffer) {
+    CARBON_DIAGNOSTIC(CppInteropFailedAccessingFileBuffer, Error,
+                      "failed accessing buffer of file '{0}': {1}", std::string,
+                      std::string);
+    emitter_.Emit(import.node_id, CppInteropFailedAccessingFileBuffer,
+                  cpp_file_path.str(), buffer.getError().message());
+    return;
+  }
```
I was expecting that you'd go the path of generating a `#include` file and letting Clang take care of these diagnostics. Given you're trying to avoid that for now, can you please take a look at whether you can use SourceBuffer::MakeFromFile here to at least reuse the diagnostic support?

bricknerb replied:
I'm using `SourceBuffer::MakeFromFile()` now, this definitely helps with avoiding code duplication.
A couple of notes:
1. From the documentation, `SourceBuffer` should be used for carbon files, and I only guess the documentation just needs to be updated.
2. `SourceBuffer` doesn't allow propagating the location where the filename was referenced so that diagnostics would point to it. I have some ideas on how to add support for that, but for now just added a TODO.

jonmeow commented on toolchain/check/testdata/interop/cpp/no_prelude/bad_import.carbon:
```diff
@@ -0,0 +1,96 @@
+
+library "[[@TEST_NAME]]";
+
+// CHECK:STDERR: fail_import_cpp.carbon:[[@LINE+4]]:1: error: imported API 'Cpp' not found [ImportNotFound]
+// CHECK:STDERR: import Cpp;
+// CHECK:STDERR: ^~~~~~
+// CHECK:STDERR:
+import Cpp;
```
Can you please make this specifically point to `Cpp` requiring a `library`, or add a TODO to that effect?

bricknerb replied:
Added a dedicated diagnostic.

jonmeow commented on toolchain/check/testdata/interop/cpp/no_prelude/bad_import.carbon:
```diff
@@ -0,0 +1,96 @@
+// CHECK:STDERR:
+import Cpp;
+
+// --- fail_import_cpp_library.carbon
+
+library "[[@TEST_NAME]]";
+
+// CHECK:STDERR: fail_import_cpp_library.carbon:[[@LINE+8]]:19: error: expected a string literal to specify the library name [ExpectedLibraryName]
+// CHECK:STDERR: import Cpp library;
+// CHECK:STDERR:                   ^
+// CHECK:STDERR:
+// CHECK:STDERR: fail_import_cpp_library.carbon:[[@LINE+4]]:19: error: semantics TODO: `handle invalid parse trees in `check`` [SemanticsTodo]
+// CHECK:STDERR: import Cpp library;
+// CHECK:STDERR:                   ^
+// CHECK:STDERR:
+import Cpp library;
```
I believe this is already tested as part of `import` parsing (vs checking); given `Cpp` doesn't do anything special with that parse tree, could this test be removed as redundant?

bricknerb replied:
Done.

jonmeow commented on toolchain/check/import_cpp.cpp:
```diff
@@ -0,0 +1,57 @@
+    // TODO: Remove the warnings part when there are no warnings.
+    CARBON_DIAGNOSTIC(
+        CppInteropParseError, Error,
+        "{0} error{0:s} and {1} warning{1:s} in C++ file '{2}':\n{3}",
```
I noticed in diagnostics that you're using single quotes. Can you please switch to backticks?

Note this is part of [diagnostic style](https://github.com/carbon-language/carbon-lang/blob/trunk/toolchain/docs/diagnostics.md#diagnostic-message-style-guide)

bricknerb replied:
Done. I've considered filenames to not be considered code but I guess we could consider them as part of the code.

jonmeow replied:
We just usually don't put filenames in the diagnostic text. :)

bricknerb committed:
Use backticks when quoting filenames.

bricknerb committed:
Make TestKindCoverage for DiagnosticKind more restrictive.

bricknerb committed:
Use SouceBuffer for reading cpp files

bricknerb committed:
Clarify multiple Cpp imports will be supported and use context.TODO to diagnose this

bricknerb committed:
Rephrase Cpp import with fuzzing error

bricknerb committed:
Improve diagnostics for bad Cpp imports

bricknerb committed:
Removed redundant `import Cpp library` test

jonmeow commented on toolchain/check/check_unit.cpp:
```diff
@@ -325,6 +334,34 @@ auto CheckUnit::ImportOtherPackages(SemIR::TypeId namespace_type_id) -> void {
+      unit_and_imports_->unit->value_stores->string_literal_values().Get(
+          import.library_id);
+
+  // TODO: Pass the import location so that diagnostics would point to it.
```
Just to note, I'd suggest focusing on multiple Cpp imports first. I think that may moot this TODO.

jonmeow commented on toolchain/check/check.cpp:
```diff
@@ -61,16 +61,39 @@ static auto RenderImportKey(ImportKey import_key) -> std::string {
+
+  if (import_package_name == CppPackageName) {
+    if (import_library_name.empty()) {
+      // Clang is not crash-resilient.
```
```suggestion
```

Looks like you copy-pasted this by accident. :)

bricknerb replied:
Oops :)

jonmeow commented on toolchain/check/check.cpp:
```diff
@@ -61,16 +61,39 @@ static auto RenderImportKey(ImportKey import_key) -> std::string {
+    if (import_library_name.empty()) {
+      // Clang is not crash-resilient.
+      CARBON_DIAGNOSTIC(CppInteropMissingLibrary, Error,
+                        "`{0}` import missing library", std::string);
+      unit_info.emitter.Emit(import.node_id, CppInteropMissingLibrary,
+                             CppPackageName.str());
```
Sorry, noticed you're doing this in a few spots. StringLiteral and std::string are supported; it's only StringRef that's disallowed because it's temporary and may not outlive the diagnostic due to buffering. CppPackageName is a StringLiteral so should work directly.
```suggestion
                        "`{0}` import missing library", llvm::StringLiteral);
      unit_info.emitter.Emit(import.node_id, CppInteropMissingLibrary,
                             CppPackageName);
```

bricknerb replied:
Changed to use `Cpp` to the format string.

jonmeow commented on toolchain/check/check.cpp:
```diff
@@ -61,16 +61,39 @@ static auto RenderImportKey(ImportKey import_key) -> std::string {
+    if (fuzzing) {
+      // Clang is not crash-resilient.
+      CARBON_DIAGNOSTIC(CppInteropFuzzing, Error,
+                        "`{0}` import found during fuzzing", std::string);
+      unit_info.emitter.Emit(import.node_id, CppInteropFuzzing,
+                             CppPackageName.str());
```
```suggestion
                        "`{0}` import found during fuzzing", llvm::StringLiteral);
      unit_info.emitter.Emit(import.node_id, CppInteropFuzzing,
                             CppPackageName);
```

bricknerb replied:
`StringLiteral` seems to not be supported, but I may be missing something here.
https://github.com/carbon-language/carbon-lang/blob/9dc450e0af9920cc5bc77d1261294978c070ae7f/toolchain/diagnostics/diagnostic.h#L137

jonmeow replied:
Oh, right. Maybe just replace `{0}` with the explicit `Cpp`?

#4416: "StringLiteral though means that it's literally written down in the toolchain, at which point it should probably be written in the format string instead of separately."

bricknerb replied:
Done.
FWIW, I wanted to avoid duplicating this string so we more easily modify it.

jonmeow commented on toolchain/check/check.cpp:
```diff
@@ -231,8 +254,10 @@ static auto BuildApiMapAndDiagnosePackaging(
       CARBON_DIAGNOSTIC(CppPackageDeclaration, Error,
-                        "`Cpp` cannot be used by a `package` declaration");
-      unit_info.emitter.Emit(packaging->names.node_id, CppPackageDeclaration);
+                        "`{0}` cannot be used by a `package` declaration",
+                        std::string);
+      unit_info.emitter.Emit(packaging->names.node_id, CppPackageDeclaration,
+                             CppPackageName.str());
```
```suggestion
                        "`{0}` cannot be used by a `package` declaration",
                        llvm::StringLiteral);
      unit_info.emitter.Emit(packaging->names.node_id, CppPackageDeclaration,
                             CppPackageName);
```

bricknerb replied:
As discussed, StringLiteral isn't supported so moved back to `Cpp`.

jonmeow reviewed (approved):
LGTM

danakj commented on toolchain/check/import_cpp.cpp:
```diff
@@ -0,0 +1,57 @@
+    // TODO: Remove the warnings part when there are no warnings.
+    CARBON_DIAGNOSTIC(
+        CppInteropParseError, Error,
+        "{0} error{0:s} and {1} warning{1:s} in C++ file `{2}`:\n{3}",
```
To Jon's point, the file name here may be incorrect, when the errors may be in other files included from it. And the clang errors will already have the filenames in them right? So maybe we can just omit the {2} here and the {1} below?

bricknerb replied:
Rephrased to refer to the Cpp import instead.
I think it's useful to quickly understand that the Cpp import triggers this.

danakj commented on toolchain/check/testdata/interop/cpp/no_prelude/bad_import.carbon:
```diff
@@ -0,0 +1,55 @@
+// CHECK:STDERR: import Cpp library "";
+// CHECK:STDERR: ^~~~~~
+// CHECK:STDERR:
+// CHECK:STDERR: "foo.h": error: error opening file for read: No such file or directory [ErrorOpeningFile]
```
Where does this `"foo.h"` come from? It's not part of the `import`

bricknerb replied:
Similar to the other use case, I believe this is related to how we use `SourceBuffer`.

danakj commented on toolchain/check/testdata/interop/cpp/no_prelude/file_not_found.carbon:
```diff
@@ -0,0 +1,23 @@
+// TIP:   bazel test //toolchain/testing:file_test --test_arg=--file_tests=toolchain/check/testdata/interop/cpp/no_prelude/file_not_found.carbon
+// TIP: To dump output, run:
+// TIP:   bazel run //toolchain/testing:file_test -- --dump_output --file_tests=toolchain/check/testdata/interop/cpp/no_prelude/file_not_found.carbon
+// CHECK:STDERR: not_found.h: error: error opening file for read: No such file or directory [ErrorOpeningFile]
```
This seems to be in the wrong place, it's not attached to the `import`?

bricknerb replied:
Yes, nice catch!
I think this is related to the bad diagnostics generated by `SourceBuffer`.
AFAIU, it puts the location of the diagnostic in the filename, which doesn't exist, so I assume this puts the diagnostic in the wrong place.
I plan to fix this by avoiding `SourceBuffer` when moving to importing C++ files by parsing a generated C++ code which includes files we want to import, though we'll probably need more iterations on making diagnostics better for C++ imports.
I hope this makes sense, please let me know otherwise and I'll work on fixing this.

danakj commented on toolchain/diagnostics/coverage_test.cpp:
```diff
@@ -42,7 +45,7 @@ constexpr DiagnosticKind UntestedDiagnosticKinds[] = {
 TEST(Coverage, DiagnosticKind) {
   Testing::TestKindCoverage(absl::GetFlag(FLAGS_testdata_manifest),
-                            R"(^ *// CHECK:STDERR: .*\.carbon:.* \[(\w+)\]$)",
+                            R"(^ *// CHECK:STDERR: .* \[(\w+)\]$)",
```
Thanks :) This looks better

bricknerb committed:
Remove bad copy paste

bricknerb committed:
Avoid StringLiteral in CARBON_DIAGNOSTIC

bricknerb committed:
Rephrase referring to Cpp import when it triggers diagnostics

danakj commented on toolchain/check/check_unit.cpp:
```diff
@@ -325,6 +334,34 @@ auto CheckUnit::ImportOtherPackages(SemIR::TypeId namespace_type_id) -> void {
+
+  // TODO: Pass the import location so that diagnostics would point to it.
+  auto source_buffer = SourceBuffer::MakeFromFile(
+      *fs_, filename, unit_and_imports_->err_tracker);
```
Could we pass `import.library_id` here to MakeFromFile, and let it default to Invalid, and when Invalid it can use the filename as the location (that's weird to me but :shrug:)?

jonmeow replied:
To be sure, I think this call should go away with the `#include` migration. I'd prefer not to change the API if this is temporary; I'd mainly suggested it in order to minimize/consolidate file error handling.

danakj replied:
Ah okay, nevermind then. We can just leave them broken for now.

bricknerb replied:
> Could we pass `import.library_id` here to MakeFromFile, and let it default to Invalid, and when Invalid it can use the filename as the location (that's weird to me but 🤷)?

FWIW, that's pretty much what I had in mind to do as a followup and why I put a TODO.
I agree though that migrating to using `#include` would make this obsolete so we should probably do that first.
